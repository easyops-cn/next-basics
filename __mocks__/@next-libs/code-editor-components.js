import React from "react";

export function CodeEditorItem() {
  return <div>code editor</div>;
}
