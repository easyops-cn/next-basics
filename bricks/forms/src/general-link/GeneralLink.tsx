import React from "react";
import { Tooltip } from "antd";
import { Link, LinkProps, GeneralIcon } from "@next-libs/basic-components";
import cssStyle from "./style.module.css";
import { isEmpty } from "lodash";
import { MenuIcon } from "@next-core/brick-types";
import classNames from "classnames";
import { FormItemWrapper, FormItemWrapperProps } from "@next-libs/forms";

export interface GeneralLinkProps
  extends Pick<LinkProps, "replace" | "target">,
    FormItemWrapperProps {
  label?: string;
  value?: string;
  text?: string;
  url?: string;
  href?: string;
  disabled?: boolean;
  tooltip?: string;
  handleClick?: () => void;
  notToJumpWhenEmpty?: boolean;
  icon?: MenuIcon;
  type?: "link" | "text";
  iconAlign?: "left" | "right";
  hideExternalIcon?: boolean;
  underLine?: boolean;
  labelColor?: string;
}

export function GeneralLink(props: GeneralLinkProps): React.ReactElement {
  const {
    text,
    url,
    href,
    tooltip,
    type,
    iconAlign = "left",
    ...linkProps
  } = props;

  const icon = props.icon ? (
    <GeneralIcon
      icon={props.icon}
      style={{
        ...(props.disabled && { color: "var(--color-disabled-text)" }),
        ...(text
          ? iconAlign === "left"
            ? { marginRight: 6 }
            : { marginLeft: 6 }
          : {}),
      }}
    />
  ) : null;

  let link;
  const labelEle = (
    <span
      className={props.underLine ? cssStyle.underLine : null}
      style={{
        borderBottomColor: props.labelColor,
        ...(props.labelColor ? { color: props.labelColor } : {}),
      }}
    >
      {text}
    </span>
  );
  const linkContent =
    iconAlign === "left" ? (
      <>
        {icon}
        {labelEle}
      </>
    ) : (
      <>
        {labelEle}
        {icon}
      </>
    );

  if (props.disabled) {
    link = <span className={cssStyle.disabledLink}>{linkContent}</span>;
  } else {
    const commonProps = {
      onClick: props.handleClick,
      className: classNames({
        [cssStyle.textLink]: type === "text",
      }),
    };
    if (props.notToJumpWhenEmpty && isEmpty(url)) {
      link = <a {...commonProps}>{linkContent}</a>;
    } else {
      link = (
        <Link to={url} href={href} {...linkProps} {...commonProps}>
          {linkContent}
        </Link>
      );
    }

    if (props.target === "_blank" && !props.hideExternalIcon) {
      link = (
        <span>
          {link}
          <span className={cssStyle.externalIcon}>
            <GeneralIcon icon={{ lib: "fa", icon: "external-link-alt" }} />
          </span>
        </span>
      );
    }
  }
  return (
    <FormItemWrapper {...props}>
      {tooltip ? <Tooltip title={tooltip}>{link}</Tooltip> : link}
    </FormItemWrapper>
  );
}
