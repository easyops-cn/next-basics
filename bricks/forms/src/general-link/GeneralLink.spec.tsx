import React from "react";
import { shallow } from "enzyme";
import { GeneralLink, GeneralLinkProps } from "./GeneralLink";
import { GeneralIcon } from "@next-libs/basic-components";

jest.spyOn(console, "warn").mockImplementation(() => void 0);

describe("GeneralLink", () => {
  it("should work", () => {
    const wrapper = shallow<GeneralLinkProps>(
      <GeneralLink
        native={true}
        label="aaa"
        url="/url"
        tooltip="tips.."
        icon={{
          lib: "antd",
          theme: "outlined",
          type: "file-search",
        }}
      />
    );

    expect(wrapper).toMatchSnapshot();
    wrapper.setProps({
      native: false,
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('target="_blank" should have external-link icon', () => {
    const wrapper = shallow<GeneralLinkProps>(
      <GeneralLink label="aaa" url="/url" target="_blank" />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should work when disabled", () => {
    const wrapper = shallow<GeneralLinkProps>(
      <GeneralLink
        label="aaa"
        disabled={true}
        icon={{
          lib: "antd",
          theme: "outlined",
          type: "file-search",
          color: "red",
        }}
      />
    );

    expect(wrapper.find("GeneralIcon").prop("style").color).toBe(
      "var(--color-disabled-text)"
    );
    expect(wrapper.find("Link").length).toBe(0);
    expect(wrapper.find("span").at(0).text()).toBe("<GeneralIcon />");
  });

  it("should work when type is text", () => {
    const wrapper = shallow<GeneralLinkProps>(
      <GeneralLink label="aaa" url="/url" type="text" />
    );
    expect(wrapper.find(".textLink").length).toBe(1);
  });

  it("should work when iconAlign is right", () => {
    const wrapper = shallow<GeneralLinkProps>(
      <GeneralLink
        label="aaa"
        url="/url"
        type="text"
        iconAlign="right"
        icon={{
          lib: "antd",
          theme: "outlined",
          type: "file-search",
        }}
        target="_blank"
        underLine={true}
        textColor={"red"}
      />
    );
    expect(wrapper.find("Link").childAt(1).props().icon).toEqual({
      lib: "antd",
      theme: "outlined",
      type: "file-search",
    });
    expect(wrapper.find(GeneralIcon).length).toBe(2);
    wrapper.setProps({
      hideExternalIcon: true,
    });
    expect(wrapper.find(GeneralIcon).length).toBe(1);
    expect(wrapper.find(".underLine").length).toBe(1);
    expect(wrapper.find(".underLine").prop("style")).toStrictEqual({
      borderBottomColor: "red",
      color: "red",
    });
  });
});
