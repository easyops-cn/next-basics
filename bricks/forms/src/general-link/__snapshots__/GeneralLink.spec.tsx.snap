// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GeneralLink should work 1`] = `
<FormItemWrapper
  icon={
    Object {
      "lib": "antd",
      "theme": "outlined",
      "type": "file-search",
    }
  }
  label="aaa"
  native={true}
  tooltip="tips.."
  url="/url"
>
  <Tooltip
    arrowPointAtCenter={false}
    autoAdjustOverflow={true}
    mouseEnterDelay={0.1}
    mouseLeaveDelay={0.1}
    placement="top"
    title="tips.."
    transitionName="zoom-big-fast"
  >
    <Link
      className=""
      icon={
        Object {
          "lib": "antd",
          "theme": "outlined",
          "type": "file-search",
        }
      }
      label="aaa"
      native={true}
      to="/url"
    >
      <GeneralIcon
        icon={
          Object {
            "lib": "antd",
            "theme": "outlined",
            "type": "file-search",
          }
        }
        style={Object {}}
      />
      <span
        className={null}
        style={
          Object {
            "borderBottomColor": undefined,
          }
        }
      />
    </Link>
  </Tooltip>
</FormItemWrapper>
`;

exports[`GeneralLink should work 2`] = `
<FormItemWrapper
  icon={
    Object {
      "lib": "antd",
      "theme": "outlined",
      "type": "file-search",
    }
  }
  label="aaa"
  native={false}
  tooltip="tips.."
  url="/url"
>
  <Tooltip
    arrowPointAtCenter={false}
    autoAdjustOverflow={true}
    mouseEnterDelay={0.1}
    mouseLeaveDelay={0.1}
    placement="top"
    title="tips.."
    transitionName="zoom-big-fast"
  >
    <Link
      className=""
      icon={
        Object {
          "lib": "antd",
          "theme": "outlined",
          "type": "file-search",
        }
      }
      label="aaa"
      native={false}
      to="/url"
    >
      <GeneralIcon
        icon={
          Object {
            "lib": "antd",
            "theme": "outlined",
            "type": "file-search",
          }
        }
        style={Object {}}
      />
      <span
        className={null}
        style={
          Object {
            "borderBottomColor": undefined,
          }
        }
      />
    </Link>
  </Tooltip>
</FormItemWrapper>
`;

exports[`GeneralLink target="_blank" should have external-link icon 1`] = `
<FormItemWrapper
  label="aaa"
  target="_blank"
  url="/url"
>
  <span>
    <Link
      className=""
      label="aaa"
      target="_blank"
      to="/url"
    >
      <span
        className={null}
        style={
          Object {
            "borderBottomColor": undefined,
          }
        }
      />
    </Link>
    <span
      className="externalIcon"
    >
      <GeneralIcon
        icon={
          Object {
            "icon": "external-link-alt",
            "lib": "fa",
          }
        }
      />
    </span>
  </span>
</FormItemWrapper>
`;
